statedb: Simplify integration with Hive

joamaki opened the pull request:
Instead of relying on the group values use an explicit registration for the tables. This allows more flexibility for how to create the tables and how to for example scope them to a module or even to keep RWTable[T] hidden inside an object without having the complicated `NewProtectedTableCell` and so on helpers which obfuscate what is going on.

```
// A constructor function for a table. Used by the real implementation and integration tests of components
// that depend on this table can use this to create and populate it.
func NewExampleTable() statedb.RWTable[*Example] {
  return statedb.NewTable[*Example]("test", ExampleIDIndex, ExampleNameIndex)
}

func example() {
  hive.New(
    statedb.Cell,
    cell.Module("example", "Example",
      cell.ProvidePrivate(NewExampleTable), // Provides RWTable[*Example] just to this module.

      // 1st way to register a table.
      cell.Invoke(statedb.RegisterTable[*Example]),

      // 2nd longer way to register a table
      cell.Invoke(func(db *statedb.DB, table statedb.RWTable[*Example]) error {
        return db.RegisterTable(table)
      }),       

    // Provide the read-only "Table[*Example]" API to the whole hive. Here or from a start
    // hook we can for example populate it before anything reads it. We get the guarantee
    // that the start hook is called before any readers if we append the hook from this constructor.
    cell.Provide(func(table statedb.RWTable[*Example]) statedb.Table[*Example] {
      return table
    }),

    // Shorter way to provide Table[*Example] using ToTable.
    // This method referenced this way has type "func(table RWTable[*Example]) Table[*Example]"
    cell.Provide(statedb.RWTable[*Example].ToTable),
  )
}

joamaki committed:
statedb: Replace cell constructors with RegisterTable

Instead of relying on the group values use an explicit registration
((*DB).RegisterTable) for the tables. This allows more flexibility
for how to create the tables and how to for example scope them to a
module or even to keep RWTable[T] hidden inside an object.

Example code:

  // A constructor function for a table. Used by the real implementation
  // and integration tests of components that depend on this table can
  // use this to create and populate it.
  func NewExampleTable() statedb.RWTable[*Example] {
    return statedb.NewTable[*Example]("test", ExampleIDIndex, ExampleNameIndex)
  }

  func example() {
    hive.New(
      statedb.Cell,
      cell.Module("example", "Example",
        cell.ProvidePrivate(NewExampleTable), // Provides RWTable[*Example] just to this module.

        // 1st way to register a table.
        cell.Invoke(statedb.RegisterTable[*Example]),

        // 2nd longer way to register a table
        cell.Invoke(func(db *statedb.DB, table statedb.RWTable[*Example]) error {
          return db.RegisterTable(table)
        }),

      // Provide the read-only "Table[*Example]" API to the whole hive. Here or from a start
      // hook we can for example populate it before anything reads it. We get the guarantee
      // that the start hook is called before any readers if we append the hook from this constructor.
      cell.Provide(func(table statedb.RWTable[*Example]) statedb.Table[*Example] {
        return table
      }),

      // Shorter way to provide Table[*Example] using ToTable.
      // This method referenced this way has type "func(table RWTable[*Example]) Table[*Example]"
      cell.Provide(statedb.RWTable[*Example].ToTable),
    )
  }

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd commented on pkg/statedb/table.go:
```diff
@@ -13,11 +13,14 @@ import (
 )
 
+// NewTable creates a new table with given name and indexes.
+// Panics if indexes are malformed which allows this to be used at
```
Is this worth the risk of a runtime panic?

As far as I can tell, it's now possible to create and register tables at an arbitrary point at runtime, and there might even be a legitimate use-case for it in the future. If that's so, then I'd much rather have to deal with an error than have a potential panic, even if it leads to slightly uglier code in the tests.

dylandreimerink replied:
That is a good point, this does invite table creation after the constructor/start hooks where panics are less ideal. Even though its not how we intend it, that is going to be inevitable.

joamaki replied:
Agreed, changed it to return error and added `MustNewTable` for the few cases where we prefer that in the tests.

bimmlerd reviewed (changes requested):
I don't like the potential `panic`s in `NewTable`.

dylandreimerink reviewed (approved):
LGTM

joamaki committed:
statedb: Return error from NewTable instead of panicing

Since with RegisterTable we can now create and register tables at runtime
it's better to avoid panics on user-data in the public API.

Added MustNewTable for tests.

Signed-off-by: Jussi Maki <jussi@isovalent.com>

bimmlerd reviewed (approved):
Thanks!
